File: ios/Momentum/ML/FeatureStore.swift
import Foundation

actor FeatureStore {
    static let shared = FeatureStore()
    
    private var metricsHistory: [DailyMetrics] = []
    private var userProfile: UserProfile {
        didSet { saveProfile() }
    }
    private let defaults = UserDefaults.standard
    private let profileKey = "momentum_user_profile"
    private let metricsKey = "momentum_metrics_history"
    
    init() {
        if let data = defaults.data(forKey: profileKey),
           let profile = try? JSONDecoder().decode(UserProfile.self, from: data) {
            userProfile = profile
        } else {
            userProfile = UserProfile()
        }
        if let data = defaults.data(forKey: metricsKey),
           let history = try? JSONDecoder().decode([DailyMetrics].self, from: data) {
            metricsHistory = history
        }
    }
    
    func ingest(_ metrics: DailyMetrics) {
        if let idx = metricsHistory.firstIndex(where: { Calendar.current.isDate($0.date, inSameDayAs: metrics.date) }) {
            metricsHistory[idx] = metrics
        } else {
            metricsHistory.append(metrics)
            userProfile.daysTracked += 1
        }
        metricsHistory.sort { $0.date < $1.date }
        saveHistory()
        updateProfileBaselines()
    }
    
    func buildVector(for date: Date) -> FeatureVector? {
        guard let metrics = metricsHistory.first(where: { Calendar.current.isDate($0.date, inSameDayAs: date) }) else {
            return nil
        }
        let calendar = Calendar.current
        let dayOfWeek = calendar.component(.weekday, from: date) - 1
        let hour = calendar.component(.hour, from: date)
        let yesterday = calendar.date(byAdding: .day, value: -1, to: date) ?? date
        let prevDayMetrics = metricsHistory.first(where: { Calendar.current.isDate($0.date, inSameDayAs: yesterday) })
        let weekMetrics = metricsHistory.filter { $0.date > calendar.date(byAdding: .day, value: -7, to: date) ?? date }
        let monthMetrics = metricsHistory.filter { $0.date > calendar.date(byAdding: .day, value: -30, to: date) ?? date }
        
        let weekSleepAvg = weekMetrics.isEmpty ? userProfile.baselineSleep : weekMetrics.map(\.sleepHours).reduce(0, +) / Double(weekMetrics.count)
        let weekStepsAvg = weekMetrics.isEmpty ? userProfile.baselineSteps : Int(weekMetrics.map(\.steps).reduce(0, +) / weekMetrics.count)
        let monthFocusAvg = monthMetrics.isEmpty ? 50 : Int(monthMetrics.map(\.focusScore).reduce(0, +) / monthMetrics.count)
        
        return FeatureVector(
            dayOfWeek: dayOfWeek,
            hour: hour,
            sleepHours: metrics.sleepHours,
            sleepQualityScore: metrics.sleepQuality.score,
            steps: metrics.steps,
            heartRateAvg: metrics.heartRateAvg,
            heartRateVariability: metrics.heartRateVariability,
            screenTimeMinutes: metrics.screenTimeMinutes,
            calendarIntensity: CalendarIntensityToInt(metrics.calendarIntensity),
            previousDayEnergy: EnergyLevelToInt(prevDayMetrics?.energyLevel ?? metrics.energyLevel),
            previousDayStress: StressLevelToInt(prevDayMetrics?.stressLevel ?? metrics.stressLevel),
            previousDayFocus: prevDayMetrics?.focusScore ?? metrics.focusScore,
            weekToDateSleepAvg: weekSleepAvg,
            weekToDateStepsAvg: weekStepsAvg,
            monthToDateFocusAvg: monthFocusAvg
        )
    }
    
    func getProfile() -> UserProfile {
        return userProfile
    }
    
    func getHistory() -> [DailyMetrics] {
        return metricsHistory
    }
    
    func recordActionResult(actionId: String, wasEffective: Bool) {
        let current = userProfile.effectiveActions[actionId] ?? 0.5
        let learningRate = 0.3
        let update = wasEffective ? current + learningRate * (1.0 - current) : current * (1.0 - learningRate)
        userProfile.effectiveActions[actionId] = min(1.0, max(0.0, update))
    }
    
    func recordTrigger(trigger: String, outcome: String) {
        let key = "\(trigger)->\(outcome)"
        userProfile.triggerPatterns[key] = (userProfile.triggerPatterns[key] ?? 0) + 1
    }
    
    func mostEffectiveAction() -> String? {
        return userProfile.effectiveActions.max(by: { $0.value < $1.value })?.key
    }
    
    func todaysMetrics() -> DailyMetrics {
        return metricsHistory.first(where: { Calendar.current.isDate($0.date, inSameDayAs: Date()) }) ?? DailyMetrics()
    }
    
    private func updateProfileBaselines() {
        let recent = metricsHistory.suffix(14)
        guard !recent.isEmpty else { return }
        userProfile.baselineSleep = recent.map(\.sleepHours).reduce(0, +) / Double(recent.count)
        userProfile.baselineSteps = Int(recent.map(\.steps).reduce(0, +) / recent.count)
        userProfile.baselineHRV = recent.map(\.heartRateVariability).reduce(0, +) / Double(recent.count)
        userProfile.learningVersion += 1
        saveProfile()
    }
    
    private func saveProfile() {
        if let data = try? JSONEncoder().encode(userProfile) {
            defaults.set(data, forKey: profileKey)
        }
    }
    
    private func saveHistory() {
        if let data = try? JSONEncoder().encode(metricsHistory) {
            defaults.set(data, forKey: metricsKey)
        }
    }
    
    private func CalendarIntensityToInt(_ intensity: CalendarIntensity) -> Int {
        switch intensity {
        case .low: return 0
        case .medium: return 1
        case .high: return 2
        case .overwhelming: return 3
        }
    }
    
    private func EnergyLevelToInt(_ level: EnergyLevel) -> Int {
        switch level {
        case .veryLow: return 0
        case .low: return 1
        case .moderate: return 2
        case .high: return 3
        case .veryHigh: return 4
        }
    }
    
    private func StressLevelToInt(_ level: StressLevel) -> Int {
        switch level {
        case .low: return 0
        case .moderate: return 1
        case .high: return 2
        case .veryHigh: return 3
        }
    }
}

File: android/app/src/main/java/com/momentum/app/ml/FeatureStore.kt
package com.momentum.app.ml

import android.content.Context
import com.momentum.app.models.CalendarIntensity
import com.momentum.app.models.DailyMetrics
import com.momentum.app.models.EnergyLevel
import com.momentum.app.models.FeatureVector
import com.momentum.app.models.SleepQuality
import com.momentum.app.models.StressLevel
import com.momentum.app.services.DataService

class FeatureStore(private val context: Context) {

    private val dataService = DataService(context)

    fun buildVector(metrics: DailyMetrics): FeatureVector {
        val history = loadMetrics()
        val weekMetrics = history.filter {
            it.date.isAfter(metrics.date.minusDays(7)) && !it.date.isAfter(metrics.date)
        }
        val monthMetrics = history.filter {
            it.date.isAfter(metrics.date.minusDays(30)) && !it.date.isAfter(metrics.date)
        }

        val weekToDateSleep = if (weekMetrics.isNotEmpty()) {
            weekMetrics.map { it.sleepHours }.average()
        } else metrics.sleepHours

        val weekToDateSteps = if (weekMetrics.isNotEmpty()) {
            weekMetrics.map { it.steps }.average().toInt()
        } else metrics.steps

        val monthToDateFocus = if (monthMetrics.isNotEmpty()) {
            monthMetrics.map { it.focusScore }.average().toInt()
        } else metrics.focusScore

        val previousDay = if (history.isNotEmpty() && history.last().date.isBefore(metrics.date)) {
            history.last()
        } else null

        return FeatureVector(
            dayOfWeek = metrics.date.dayOfWeek.value,
            hour = java.util.Calendar.getInstance().get(java.util.Calendar.HOUR_OF_DAY),
            sleepHours = metrics.sleepHours,
            sleepQualityScore = metrics.sleepQuality.score,
            steps = metrics.steps,
            heartRateAvg = metrics.heartRateAvg,
            heartRateVariability = metrics.heartRateVariability,
            screenTimeMinutes = metrics.screenTimeMinutes,
            calendarIntensity = metrics.calendarIntensity.ordinal,
            previousDayEnergy = previousDay?.energyLevel?.ordinal ?: 2,
            previousDayStress = previousDay?.stressLevel?.ordinal ?: 1,
            previousDayFocus = previousDay?.focusScore ?: 50,
            weekToDateSleepAvg = weekToDateSleep,
            weekToDateStepsAvg = weekToDateSteps,
            monthToDateFocusAvg = monthToDateFocus
        )
    }

    fun ingest(metrics: DailyMetrics) {
        dataService.saveMetrics(metrics)
    }

    fun loadMetrics(): List<DailyMetrics> {
        return dataService.loadAllMetrics().sortedBy { it.date }
    }

    fun loadMetricsForDate(date: java.time.LocalDate): DailyMetrics? {
        return dataService.loadMetrics(date)
    }

    fun recordActionResult(actionId: String, effectiveness: Double) {
        val actionsKey = "action_effectiveness"
        val existing = dataService.getString(actionsKey, "{}")
        val map = mutableMapOf<String, Double>()
        try {
            val entries = existing.removeSurrounding("{", "}")
                .split(",")
                .filter { it.isNotBlank() }
            entries.forEach { entry ->
                val parts = entry.split("=")
                if (parts.size == 2) {
                    map[parts[0].trim()] = parts[1].trim().toDouble()
                }
            }
        } catch (_: Exception) {
        }
        val current = map[actionId] ?: 0.0
        map[actionId] = (current + effectiveness) / 2.0
        val serialized = map.entries.joinToString(",") { "${it.key}=${it.value}" }
        dataService.saveString(actionsKey, "{$serialized}")
    }

    fun getActionEffectiveness(): Map<String, Double> {
        val key = "action_effectiveness"
        val existing = dataService.getString(key, "{}")
        val map = mutableMapOf<String, Double>()
        try {
            val entries = existing.removeSurrounding("{", "}")
                .split(",")
                .filter { it.isNotBlank() }
            entries.forEach { entry ->
                val parts = entry.split("=")
                if (parts.size == 2) {
                    map[parts[0].trim()] = parts[1].trim().toDouble()
                }
            }
        } catch (_: Exception) {
        }
        return map
    }
}
